pyccl/cls.py

```diff
@@ -27,6 +27,9 @@ def angular_cl(cosmo, cltracer1, cltracer2, ell, p_of_k_a=None,
             :math:`C_\\ell`, for the pair of tracers, as a function of
             :math:`\\ell`.
     """
+    # we need the distances for the integrals
+    cosmo.compute_distances()
```

Comment:
So it is gonna be kind of annoying if there's no way for the cosmo object to realize that it needs to compute stuff at the C level. Can you clarify why this is necessary?

Reply:
With threading at the c level it is basically impossible to properly sync the global state in the cosmology object. By moving all of that to python, we can avoid this. Additionally future dev will require that all global state be initialized at the python level before calling into C. So this won’t be annoying as much as just the way things work. There are right now a few exceptions to this in the code. These will be removed over time.

Reply:
To be a bit more clear, the idea is that when coding in C with threads, you never have to worry about the cosmology state being change by another thread. You do have to test for the data, but if it is not there you simply set an error and move on.

Reply:
ok, I'm sure you're right. Let me look at the C code and I'll get back to this one.

Reply:
I did some experimenting with allowing this to happen. The issue comes when you have to sync the data. Basically you need a mechanism to 1) lock the Cosmo structure so you can modify it and 2) block the execution of other threads until the data is updated. I am sure we could work out how to do this, but it would be complicated and bug prone. By forcing the init to only happen from python when the code is single threaded, we effectively use the GIL to take care of both of these things. It also makes the programming model a lot simpler.